src/psyclone/dynamo0p3.py

```diff
@@ -311,29 +318,37 @@ def qr_basis_alloc_args(first_dim, basis_fn):
             "dynamo0p3.qr_basis_alloc_args(). Should be one of: "
             "{1}".format(basis_fn["shape"], VALID_QUADRATURE_SHAPES))
 
+    if basis_fn["qr_var"]:
+        qr_var = "_"+basis_fn["qr_var"]
+    else:
+        # We have no name for the QR variable (as will be the case when
+        # generating a kernel stub) so invent one.
+        qr_var = "_qr_" + basis_fn["shape"].split("_")[-1]
+
     # Dimensionality of the basis arrays depends on the
     # type of quadrature...
     # if basis_fn["shape"] == "gh_quadrature_xyz":
     #     alloc_args = [first_dim, get_fs_ndf_name(basis_fn["fspace"]),
     #          "np_xyz"+"_"+basis_fn["qr_var"]]
     if basis_fn["shape"] == "gh_quadrature_xyoz":
-        alloc_args = [first_dim, get_fs_ndf_name(basis_fn["fspace"])]
-        if basis_fn["qr_var"]:
-            alloc_args += ["np_xy"+"_"+basis_fn["qr_var"],
-                           "np_z"+"_"+basis_fn["qr_var"]]
-        else:
-            # We have no name for the QR variable (as will be the case when
-            # generating a kernel stub).
-            alloc_args += ["np_xy", "np_z"]
+        alloc_args = [first_dim, get_fs_ndf_name(basis_fn["fspace"]),
+                      "np_xy"+qr_var, "np_z"+qr_var]
     # elif basis_fn["shape"] == "gh_quadrature_xoyoz":
     #     alloc_args = [first_dim, get_fs_ndf_name(basis_fn["fspace"]),
     #                   "np_x"+"_"+basis_fn["qr_var"],
     #                   "np_y"+"_"+basis_fn["qr_var"],
     #                   "np_z"+"_"+basis_fn["qr_var"]]
+    elif basis_fn["shape"] == "gh_quadrature_face":
+        alloc_args = [first_dim, get_fs_ndf_name(basis_fn["fspace"]),
+                      "np_xyz"+qr_var, "nfaces"+qr_var]
+    elif basis_fn["shape"] == "gh_quadrature_edge":
+        alloc_args = [first_dim, get_fs_ndf_name(basis_fn["fspace"]),
+                      "np_xyz"+qr_var, "nedges"+qr_var]
     else:
         raise NotImplementedError(
-            "Quadrature shape '{0}' not yet supported in "
-            "dynamo0p3.qr_basis_alloc_args().".format(basis_fn["shape"]))
+            "unrecognised shape ({0}) specified in "
+            "dynamo0p3.qr_basis_alloc_args(). Should be one of: "
+            "{1}".format(basis_fn["shape"], VALID_QUADRATURE_SHAPES))
```

Comment:
VALID_QUADRATURE_SHAPES is a list so it doesn't really make sense to put quotes around it (and I think that the members of the list will automatically be quoted anyway because they're strings).